File: test_drive/lib/widgets/hybrid_editor/core/transactions/operation.dart
import '../document.dart' show Document;

/// Types of operations that can be performed on the document
enum OperationType {
  insertText,
  deleteText,
  insertBlock,
  deleteBlock,
  updateBlock,
  replaceBlocks,
  setSelection,
  updateListLevel,
  composite, // Multiple operations grouped together
}

/// Represents a single atomic operation on the document.
/// 
/// Operations are reversible, allowing for undo/redo functionality.
abstract class Operation {
  OperationType get type;
  
  /// Applies the operation to the document and returns the new document
  Document apply(Document document);
  
  /// Returns the inverse operation for undo
  Operation inverse(Document document);
}

File: components/editor/core/operations/Operation.ts
import type { Document } from "../Document";

/// Types of operations that can be performed on the document
export enum OperationType {
	insertText = "insertText",
	deleteText = "deleteText",
	insertBlock = "insertBlock",
	deleteBlock = "deleteBlock",
	updateBlock = "updateBlock",
	updateBlockAttributes = "updateBlockAttributes",
	replaceBlocks = "replaceBlocks",
	setSelection = "setSelection",
	composite = "composite",
}

/// Represents a single atomic operation on the document.
///
/// Operations are reversible, allowing for undo/redo functionality.
export interface Operation {
	readonly type: OperationType;

	/// Applies the operation to the document and returns the new document
	apply(document: Document): Document;

	/// Returns the inverse operation for undo
	inverse(document: Document): Operation;
}
